src/promptflow-core/promptflow/core/_prompty_utils.py

```diff
@@ -653,11 +653,17 @@ def try_parse_tool_call_id_and_content(role_prompt):
 def try_parse_tool_calls(role_prompt):
     # customer can add ## in front of tool_calls for markdown highlight.
     # and we still support tool_calls without ## prefix for backward compatibility.
-    pattern = r"\n*#{0,2}\s*tool_calls\s*:\s*\n+\s*(\[.*?\])"
+
+    # Previously used pattern (commented) cannot parse tool_calls with internal square brackets
+    # common with function call arguments. Updated pattern matches content before tool_calls array
+    # and removes it before parsing the array.
+    # pattern = r"\n*#{0,2}\s*tool_calls\s*:\s*\n+\s*(\[.*?\])"
+    pattern = r"(\n*#{0,2}\s*tool_calls\s*:\s*\n+\s*)\["
```

Comment:
please fix the tool test, https://github.com/microsoft/promptflow/actions/runs/9543286774/job/26300487985?pr=3389

Reply:
Working in latest commit. I had to implement bracket counting, which I was hoping to avoid, to make this work since this test results in parsing of an assistant tool_calls message that has excess content due to an incorrect role definition. The previous implementation assumed the `role_prompt` would be parsed with `]` as the last non-whitespace character which cannot be assumed based on this test case.

I also added additional tests that would fail under the previous implementation to help guide any future changes.